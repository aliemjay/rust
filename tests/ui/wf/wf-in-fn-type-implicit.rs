// issue: #104005

// Function type parameters with implicit type annotations were not
// checked for well-formedness, which allowed incorrect borrowing.

use std::fmt::Display;

trait Displayable {
    fn display(self) -> Box<dyn Display>;
}

impl<T: Display> Displayable for (T, Option<&'static T>) {
    fn display(self) -> Box<dyn Display> {
        Box::new(self.0)
    }
}

fn extend_lt<T, U>(val: T) -> Box<dyn Display>
where
    (T, Option<U>): Displayable,
{
    Displayable::display((val, None))
}

fn main() {
    // This used to compile
    let val = extend_lt(&String::from("blah blah blah"));
    //~^ ERROR temporary value dropped while borrowed
    println!("{}", val);
}
