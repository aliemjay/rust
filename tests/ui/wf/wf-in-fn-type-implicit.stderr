error[E0716]: temporary value dropped while borrowed
  --> $DIR/wf-in-fn-type-implicit.rs:27:26
   |
LL |     let val = extend_lt(&String::from("blah blah blah"));
   |               -----------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^-- temporary value is freed at the end of this statement
   |               |          |
   |               |          creates a temporary value which is freed while still in use
   |               argument requires that borrow lasts for `'static`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0716`.
